_posts/2021-03-09-bezier-fitting.md

```diff
@@ -0,0 +1,196 @@
++ \ & 50 \delta_1 \sin \theta_1 \\
++ \ & 15 \delta_0^2 \sin \theta_0 \cos \theta_0 \\
+- \ & 15 \delta_1^2 \sin \theta_1 \cos \theta_1 \\
+- \ & \delta_0 \delta_1 (33\sin \theta_0 \cos \theta_1 + 9 \cos \theta_0 \sin \theta_1) \\
+- \ &9 \delta_0^2 \delta_1 \sin(\theta_0 + \theta_1) \cos \theta_0 \\
++ \ & 9 \delta_0 \delta_1^2 \sin(\theta_0 + \theta_1) \cos \theta_1
+) / 280
+\end{align}
+$$
+
+It's perfectly reasonable to use numerical techniques to solve this equation (it's much cheaper than computing an error norm relative to the original curve), but again, we can do better.
+
+### A quartic polynomial
+
+We're looking for the simultaneous solution of the area and x-moment constraints. Using some mathematical tricks (very similar to those used in the Penner paper), we can reduce the whole thing to a single quartic polynomial.
+
+* Transform instances of $\delta_0 \delta_1$ into an equation of the form $a\delta_0 + b\delta_1 + c$
+
+* Replace all instances of $\delta_1$ by its area-preserving solution in terms of $\delta_0$.
+
+* Multiply the polynomial by $(\delta_0 \sin (\theta_0 + \theta_1) - 2 \sin \theta_1)^2$
+
+Given a [quartic polynomial], there are both analytical and numeric solutions (the latter can still be desirable because it can be tricky to achieve numerical robustness).
+
+In addition to the increased leverage for solving techniques, the existence of a polynomial gives us great insight into the solution space, for example the number of solutions.
+
+### The fourth solution
+
+A quartic polynomial can have up to four solutions, but so far we've been talking about triplets of solutions. What about the fourth?
+
+It turns out that it's also a valid solution of the area and x-moment constraints, but not at all close to the original curve, as it contains a loop. When computing signed area, the lobe enclosed by the loop has opposite sign as the rest of the curve, and similarly for moment.
+
+[TODO show example]
+
+Thus, in the general case we can talk about a triplet of similar curves and one "nemesis" that has a different loop structure, all of which share the same area and x-moment.
+
+## Near misses
+
+If we only consider exact solutions of the x-moment constraint, we may miss some error minima. Consider the Euler spiral with deflections of 0.285 and 0.35 radians as a source curve to fit. The graph of x-moment and error looks like this:
+
+![Graph of x-moment showing near approach to axis](/assets/bezier_fit_near_miss.png)
+
+Here, on the right side of the graph, the x-moment approaches zero but doesn't actually cross the x axis.
+
+* Optimal solution is actually near x-moment minimum.
+
+* These solutions correspond to "surprising oval" in Penner paper.
+
+## Conclusion and discussion
+
+* Cubic beziers have O(n^6) scaling. This algorithm lets you realize that level of accuracy.
+
+* In addition to a very fast algorithm, insights into cubic beziers. The presence of triplets of C-shaped curves.
```

Comment:
I'd probably call this a family at this point, rather than triplets: the three curves can look different enough to feel like "different curves", but they're all solutions to the same problem.

Reply:
The point is that they don't look different at all. All of them are within the same error bound as an O(n^4)-scaling q&d solution like Potrace. Hopefully showing examples will make this clearer.